### PR TITLE
Fix SM2 signature to follow the specification

### PR DESCRIPTION
**Pull request by AlverLyu**
Before SM2 signing and verifying, combine the signed message with user
id, curve parameters and public key, as described in SM2 specification.

### CODE REVIEW

File: crypto/sm2/sm2.go [hunk 1]
```diff
@@ -18,7 +18,8 @@ type zr struct {
 }
 
 const (
-	aesIV = "IV for <SM2> CTR"
+	aesIV  = "IV for <SM2> CTR"
+	USERID = "1234567812345678"
```
Comment (dreamfly281) [inline crypto/sm2/sm2.go]: Does the value be defined by the standard?

Comment (AlverLyu) [inline crypto/sm2/sm2.go]: It is the default value, and used in all the SM2 CA currently.

Comment (dreamfly281) [inline crypto/sm2/sm2.go]: good to know that

File: crypto/sm2/sm2.go [hunk 2]
```diff
@@ -171,9 +173,29 @@ func Verify(algSet *util.CryptoAlgSet, publicKeyX *big.Int, publicKeyY *big.Int,
...
+func combineZ(raw []byte, publicKeyX, publicKeyY *big.Int, userID string) []byte {
+	h := sm3.New()
+
+	id := []byte(userID)
+	len := len(id) * 8
+	blen := []byte{byte((len >> 8) & 0xff), byte(len & 0xff)}
+
+	h.Write(blen)
+	h.Write(id)
+	h.Write([]byte("\xFF\xFF\xFF\xFE\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\x00\x00\x00\x00\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFC"))
```
Comment (d5c5ceb0) [inline crypto/sm2/sm2.go]: How about defining a const variable “A" to replace the magic number?

Comment (AlverLyu) [inline crypto/sm2/sm2.go]: Reasonable, though it is only used here currently.

Comment (dreamfly281) [inline crypto/sm2/sm2.go]: please update the commit based on the comments and resubmit the code

**Pull request merged**